### PR TITLE
Add proof skolem cache

### PR DESCRIPTION
**Pull request by ajreynol**
This adds the general utility for maintaining mappings from Skolems to their witness form via attributes.

I am sending this as a PR now since it would be helpful to use this class for fixing some of the recent unconstrained-simp bugs.

### CODE REVIEW

File: src/expr/proof_skolem_cache.h [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+#include <string>
+
+#include "expr/node.h"
+
+namespace CVC4 {
+
+/**
+ * A manager for skolems that can be used in proofs. This is designed to be
+ * trusted interface to NodeManager::mkSkolem, where one
```
Comment (HanielB) [inline src/expr/proof_skolem_cache.h:28]: to be a

File: src/expr/proof_skolem_cache.h [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+   * @param flags The flags for the Skolem (see NodeManager::mkSkolem)
+   * @return The skolem whose witness form is registered by this class.
+   */
+  static Node mkSkolem(Node v,
+                       Node pred,
+                       const std::string& prefix,
+                       const std::string& comment = "",
+                       int flags = NodeManager::SKOLEM_DEFAULT);
+  /**
+   * Same as above, but for special case for (witness ((x T)) (= x t))
```
Comment (HanielB) [inline src/expr/proof_skolem_cache.h:92]: phrasing

File: src/expr/proof_skolem_cache.cpp [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                                int flags)
+{
+  Assert(v.getKind() == BOUND_VARIABLE);
+  // make the witness term
+  NodeManager* nm = NodeManager::currentNM();
+  Node bvl = nm->mkNode(BOUND_VAR_LIST, v);
+  // translate pred to witness form, since pred itself may contain skolem
+  Node predw = getWitnessForm(pred);
+  // make the witness term, which should not contain any skolem
+  Node w = nm->mkNode(CHOICE, bvl, predw);  // will change to WITNESS
```
Comment (HanielB) [inline src/expr/proof_skolem_cache.cpp:51]: Hasn't master already been updated to use `WITNESS` as the kind?

File: src/expr/proof_skolem_cache.h [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+   * where T is the type of t. This skolem is unique for each t.
+   */
+  static Node mkPurifySkolem(Node t,
```
Comment (HanielB) [inline src/expr/proof_skolem_cache.h:95]: I think it'd be good to say here why there is a dedicated attribute for this case.

**Comment by ajreynol** [comment]
Marking do not merge since I have at least one more feature I'd like to add to this class on stringsPf branch.

File: src/expr/proof_skolem_cache.cpp (hunk 3 above)
Comment (ajreynol) [reply src/expr/proof_skolem_cache.cpp:51]: Not yet.

**Comment by ajreynol** [comment]
@HanielB  I addressed the review. Notice that I also added a few more features in my latest commit that are being used on the strings proof branch.

File: src/expr/proof_skolem_cache.h (hunk 4 above)
Comment (HanielB) [reply src/expr/proof_skolem_cache.h:110]: You didn't address this comment.

**Pull request merged**